packages/fxa-settings/src/lib/hooks/useValidate/index.tsx

```diff
@@ -0,0 +1,29 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import {
+  ModelDataProvider,
+  ModelDataStore,
+  UrlQueryData,
+} from '../../model-data';
+import { ReachRouterWindow } from '../../window';
+
+let urlQueryDataInstance: UrlQueryData;
+
+// TODO: We should be able to access which param has errored so we can handle
+// them individually if we need to.
+export function useValidatedQueryParams<T extends ModelDataProvider>(
+  QueryParamModel: new (modelData: ModelDataStore) => T
+) {
+  if (!urlQueryDataInstance) {
+    urlQueryDataInstance = new UrlQueryData(new ReachRouterWindow());
```

Comment:
It might be better to avoid this. There's a _chance_ this pattern would create a lot of garbage collection.